### ISSUE TITLE
Support local audio files

### ISSUE THREAD

**Issue by brnwlshubh** [z-huang/InnerTune#3]
Local music player support for any audio files saved in the phone.

**Comment by z-huang** [comment]
This is the main problem when I started to make this project: whether to import local media files to the library or not? Currently, I don't plan to introduce this feature for these aspects below.

1. User Experience
To show a song's information clearly, we need at least the name, artist, and if better, the artwork. However, they should be accessed through the ID3 metadata of a local song, which is often not set properly. Though people may put the information in the filename, the format varies. It's difficult to analyze song names and artists automatically. Therefore, songs may lack artist names and artwork, and look ugly. If all songs are added from YouTube (Music), they will have an initial metadata attached and look pretty.

2. Consistency
Actually, almost every song is available in YouTube or YouTube Music, so there won't be a problem that you can't find a song in the app. Also, many streaming services such as Spotify, KKBOX, MyMusic don't let you import local songs. You should discover and add them by yourself. In addition, there won't be the problem of duplicate songs. Songs from a maintained source should be better and cause less problems.

3. Implementation Difficulty
It's hard to manage songs from YouTube and in local storage together. Since the source is different, operations such as download, delete, edit can be confusing and not unified. Also, if the app is designed this way, users might well download music from other places and use a music player to play all local songs.

In conclusion, the goal of this app is to provide a free streaming service with downloadable contents. There are features like exporting songs through Android SAF planned to be implemented to enhance convenience.
If you have any thoughts about an ideal concept for this app, don't hesitate to tell me and I'll put it into consideration. 😀

**Comment by z-huang** [comment]
Yes, I'm planning to do so, but it can take a considerable amount of time since NewPipeExtractor doesn't provide the YouTube Music kiosk extractor. Now the explore tab in the app is lack of content and recommendations are expected to be put there. There are some libraries regarding YouTube Music API written in python, and I'll dig into them and try to implement the feature in Kotlin.

**Comment by z-huang** [comment]
As #43 has a more descriptive title, I'll close this one.

**Comment by noClaps** [comment]
> Actually, almost every song is available in YouTube or YouTube Music

There's some songs that aren't available on YT Music or YouTube that are available on other platforms. An example of this would be songs from the album Pebble Brain by Lovejoy.

> Also, many streaming services such as Spotify

Most popular streaming services allow you to import local songs. I know for a fact that both Spotify, YT Music, and Deezer allow this, and I'm pretty sure others do too.

> It's hard to manage songs from YouTube and in local storage together

Why not have a separate section for local music? YT Music does this, for example.

> Therefore, songs may lack artist names and artwork, and look ugly.

That should be up to the user to get proper metadata for their files. This would affect any local music player as well, if the songs weren't tagged properly. Optionally, you could provide a tag editor, though that probably isn't necessary for a streaming app such as this.

> Since the source is different, operations such as download, delete, edit can be confusing and not unified.

Maybe you can mark the local songs as downloaded? As for delete, I think the delete option for local songs should delete the song file from the device. I think Retro Music does this. Otherwise, you can leave it up to the user to delete the song file manually, like how Auxio does.

> In addition, there won't be the problem of duplicate songs.

Probably should be up to the user to not have duplicates in their library. If you want, you can add a source tag for each song, if it came from local storage or from YT Music.

> if the app is designed this way, users might well download music from other places and use a music player to play all local songs.

We could, but one of the benefits of using this app is that I can stream and not have to carry around gigabytes of music files to every device I use, and I can add and remove songs quickly without having to delete the music file, and download new music and then transfer it to my phone.

**Comment by z-huang** [comment]
Okay, you convinced me that local audio files should be supported. I'll reopen this issue.
I haven't figured out how to design the database to support it. When I'm done with `feature/innertube` and other trivial bugs or enhancement, I'll go back on this.

**Comment by z-huang** [comment]
Not planned because it's too "[advanced](https://github.com/z-huang/InnerTune/discussions/421#discussioncomment-6198050)".

**Comment by AwkMedia** [comment]
> Not planned because it's too "[advanced](https://github.com/z-huang/InnerTune/discussions/421#discussioncomment-6198050)".

Is this the kind of advanced where you would consider accepting a PR?

**Comment by z-huang** [comment]
Yes, PR is welcome.
